### PR TITLE
Sdem periodic backward coupling

### PR DESCRIPTION
**Pull request by GuillermoCasas**
**📝 Description**
Making the backward projection compatible with periodic domains. There remains some work to do:

1. Unifying the configure files (there is considerable amount of copied code)
2. Simplifying the search algorithm (there is room for that but requires careful investigations)
3. Using the `IntersectionBox` function from the configure in the `FillObjectPeriodic` function. Now there is much redundancy. 

**🆕 Changelog**
- Changed the search class to be used in binbased for the homogenisation method (now it is the same as the one used in DEM)
- Recycled the long unused node configure to search fluid nodes from the DEM nodes
- Added two tests, one to DEM and one to SwimmingDEM:
    - The one in DEM checks that the right nodes are found by two spheres in a periodic domain.
    - The one in SwimmingDEM checks that the application of this periodic projection (as opposed to a projection that does not take into account periodicity) leads to a smaller number of nonlinear iterations of the fluid solver.

### CODE REVIEW

File: applications/DEMApplication/custom_search/search_structure_periodic.h [hunk 1]
```diff
@@ -88,7 +90,7 @@ class SubBinAxisPeriodic
             return static_cast<SizeType>(Max-Min);
         }
         else {
-            return static_cast<SizeType>(MaxSize - 1 + (Max - Min));
+            return static_cast<SizeType>(MaxSize + 1 + Max - Min);
```
Comment (GuillermoCasas) [inline applications/DEMApplication/custom_search/search_structure_periodic.h:93]: This was an error! When I fixed this, my test started to work

File: applications/DEMApplication/custom_search/bins_dynamic_objects_periodic.h [hunk 2]
```diff
@@ -342,8 +365,8 @@ inline void NextIndex(IndexType& Index, int& box_size_counter, PointType& MinCel
     }
     else {
         Index = 0;
-        MinCell[dimension] += - MinCell[dimension];
-        MaxCell[dimension] += - MaxCell[dimension] + this->mCellSize[dimension];
+        MinCell[dimension] = 0;
```
Comment (maceligueta) [inline applications/DEMApplication/custom_search/bins_dynamic_objects_periodic.h:368]: This looks kinda dangerous. Isn't it changing the behaviour of the search?

File: applications/DEMApplication/custom_utilities/discrete_particle_configure.h [hunk 3]
```diff
@@ -231,7 +226,7 @@ class DiscreteParticleConfigure
         SphericParticle* p_particle1 = static_cast<SphericParticle*>(&*rObj_1);
         SphericParticle* p_particle2 = static_cast<SphericParticle*>(&*rObj_2);
         const double radius_sum      = p_particle1->GetSearchRadius() + p_particle2->GetSearchRadius();
-        const bool intersect         = floatle((distance_2 - radius_sum * radius_sum),0);
+        const bool intersect         = floatle(distance_2, std::pow(radius_sum, 2));
```
Comment (maceligueta) [inline applications/DEMApplication/custom_utilities/discrete_particle_configure.h:229]: pow is slower with some compilers. We could leave radius_sum*radius_sum

File: applications/DEMApplication/custom_utilities/node_configure.h [hunk 4]
```diff
@@ -212,21 +274,40 @@ class NodeConfigure
...
+    }
+
     static inline bool floateq(double a, double b) {
         return std::fabs(a - b) < std::numeric_limits<double>::epsilon();
     }
-    
+
     static inline bool floatle(double a, double b) {
-        return std::fabs(a - b) < std::numeric_limits<double>::epsilon() || a < b;
+        return a < b || std::fabs(a - b) < std::numeric_limits<double>::epsilon();
```
Comment (maceligueta) [inline applications/DEMApplication/custom_utilities/node_configure.h:304]: nice

File: applications/DEMApplication/custom_search/bins_dynamic_objects_periodic.h (hunk 2 above)
Comment (GuillermoCasas) [reply applications/DEMApplication/custom_search/bins_dynamic_objects_periodic.h:368]: We were subtracting the LHS to the LHS. The result is 0

File: applications/DEMApplication/custom_utilities/discrete_particle_configure.h (hunk 3 above)
Comment (GuillermoCasas) [reply applications/DEMApplication/custom_utilities/discrete_particle_configure.h:229]: Really? I thought this was well taken care of by the compiler. I would prefer leaving it as pow; it is cleaner and will only get more optimized in time.

Comment (maceligueta) [reply applications/DEMApplication/custom_utilities/discrete_particle_configure.h:229]: I forgot to paste the comparison study: https://stackoverflow.com/questions/2940367/what-is-more-efficient-using-pow-to-square-or-just-multiply-it-with-itself

Comment (GuillermoCasas) [reply applications/DEMApplication/custom_utilities/discrete_particle_configure.h:229]: My conclusion from the comparison is that indeed the difference is not really significant for 2. So I stick to my previous comments.

**Review by maceligueta — APPROVED** [review]

**Pull request merged**